pulsar-client-cpp/python/src/authentication.cc

```diff
@@ -32,6 +32,59 @@ struct AuthenticationTlsWrapper : public AuthenticationWrapper {
     }
 };
 
+struct TokenSupplierWrapper {
+    PyObject* _pySupplier;
+
+    TokenSupplierWrapper(py::object pySupplier) :
+        _pySupplier(pySupplier.ptr()) {
+        Py_XINCREF(_pySupplier);
+    }
+
+    TokenSupplierWrapper(const TokenSupplierWrapper& other) {
+        _pySupplier= other._pySupplier;
+        Py_XINCREF(_pySupplier);
+    }
+
+    TokenSupplierWrapper& operator=(const TokenSupplierWrapper& other) {
+        _pySupplier = other._pySupplier;
+        Py_XINCREF(_pySupplier);
+        return *this;
+    }
+
+    virtual ~TokenSupplierWrapper() {
+        Py_XDECREF(_pySupplier);
+    }
+
+    std::string operator()() {
+        PyGILState_STATE state = PyGILState_Ensure();
```

Comment:
why do you need to acquire the GIL? _pySupplier is immutable.

Reply:
We need the GIL because we're calling into python code (supplied by the user)

Reply:
:+1: